pvlib/modelchain.py

```diff
@@ -972,7 +1140,88 @@ def run_model(self, weather, times=None):
         self.aoi_model()
         self.spectral_model()
         self.effective_irradiance_model()
-        self.temperature_model()
+
+        self.run_model_from_effective_irradiance(weather)
+
+        return self
+
+    def run_model_from_poa(self, data):
+        """
+        Run the model starting with broadband irradiance in the plane of array.
+
+        Data must include direct, diffuse and total irradiance (W/m2) in the
+        plane of array. Reflections and spectral adjustments are made to
+        calculate effective irradiance (W/m2).
+
+        This method calculates:
+            * effective irradiance
+            * cell temperature
+            * DC output
+            * AC output
+
+        Parameters
+        ----------
+        data : DataFrame
+            Required column names must include ``'poa_global'``,
+            ``'poa_direct'`` and ``'poa_diffuse'``. If optional columns
+            ``'temp_air'`` and ``'wind_speed'`` are not provided, air
+            temperature of 20 C and wind speed of 0 m/s are assumed.
+            If optional column ``'cell_temperature'`` is provided, these values
+            are used instead of `temperature_model`. If optional column
+            `module_temperature` is provided, `temperature_model` must be
+            ``'sapm'``.
+
+        Returns
+        -------
+        self
+
+        Assigns attributes: ``weather``, ``total_irrad``, ``solar_position``,
+        ``airmass``, ``aoi`` and ``effective_irradiance``.
+        Calls `run_model_from_effective_irradiance`
+        to assign ``cell_temperature``, ``dc``, ``ac``,
+        ``losses``, ``diode_params`` (if dc_model is a single diode model)
+        """
+
+        self.prepare_inputs_from_poa(data)
+
+        self.aoi_model()
+        self.spectral_model()
+        self.effective_irradiance_model()
+
+        self.run_model_from_effective_irradiance(data)
+
+        return self
+
+    def run_model_from_effective_irradiance(self, data=None):
```

Comment:
does this method work as a standalone entry point? It appears we need something like `prepare_inputs_from_effective_irradiance`.

Reply:
It doesn't stand alone in this iteration. I wanted `run_model_from_effective_irradiance` to encapsulate the temperature/DC/AC/losses models, and use this function at the end of e.g. `run_model_from_poa` to avoid duplicate code. Doing so means that `run_model_from_effective_irradiance` can't set `ModelChain.total_irrad` for example.

I think a way forward is to make the current function private and make a separate public function 1) sets weather, total_irrad etc. then 2) runs the model forward from effective_irradiance.